src/mixturemodels.jl

```diff
@@ -0,0 +1,46 @@
+function _mixlogpdf1(d::AbstractMixtureModel, x)
+    # using the formula below for numerical stability
+    #
+    # logpdf(d, x) = log(sum_i pri[i] * pdf(cs[i], x))
+    #              = log(sum_i pri[i] * exp(logpdf(cs[i], x)))
+    #              = log(sum_i exp(logpri[i] + logpdf(cs[i], x)))
+    #              = m + log(sum_i exp(logpri[i] + logpdf(cs[i], x) - m))
+    #
+    #  m is chosen to be the maximum of logpri[i] + logpdf(cs[i], x)
+    #  such that the argument of exp is in a reasonable range
+    #
+
+    K = ncomponents(d)
+    p = probs(d)
+    # use Buffer to avoid mutating arrays.
+    # lp = Vector{eltype(p)}(undef, K)
+    lp = Zygote.Buffer(p, K)
+    m = -Inf   # m <- the maximum of log(p(cs[i], x)) + log(pri[i])
+    @inbounds for i in eachindex(p)
```

Comment:
Try defining a non-mutating function that takes in `p` as input and returns `lp` as output. Use a `map` block. `m` is just `maximum(lp)`.

Reply:
For example,
```julia
map(eachindex(p)) do i
   ...
end
```

Reply:
jus to be clear, you are fine with using Zygote.Buffer, which means we need to add Zygote to Project.toml.

Your comments are only about refactoring the for loop right ?